src/cockpit/packages.py

```diff
@@ -15,193 +15,204 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-
 import collections
+import functools
 import gzip
-import hashlib
 import json
 import logging
 import mimetypes
 import os
-import re
 import shutil
 from pathlib import Path
-from typing import Callable, ClassVar, Dict, List, Optional, Pattern, Tuple
+from typing import Any, Callable, ClassVar, Dict, Iterable, List, Optional, Set, Tuple, TypeVar, Union
 
 from cockpit._vendor.systemd_ctypes import bus
 
 from . import config
+from ._version import __version__
 
-VERSION = '300'
 logger = logging.getLogger(__name__)
 
+JSON = Union[Dict[str, 'JSON'], List['JSON'], str, int, float, bool, None]
+Object = Dict[str, JSON]
 
-# An entity to serve is a set of bytes plus a content-type/content-encoding pair
-Entity = Tuple[bytes, Tuple[Optional[str], Optional[str]]]
-Entities = Dict[str, Entity]
+# HACK: Type narrowing over Union types is not supported in the general case,
+# but this works for the case we care about: knowing that when we pass in an
+# Object, we'll get an Object back.
+J = TypeVar('J', Object, JSON)
```

Comment:
hmm, single-letter type vars? Could this be `JSONObject`?

Reply:
The convention here is single characters...

Reply:
Hmmkay. Where is this being used? I don't see `-> J` nor `: J` anywhere. I also cycled through searching or 'J' in the diff, and didn't see anything.

Reply:
Hmm, is it? This is being used as `-> J` or `: J`, and types are usually readable names...

Reply:
Take a look at https://docs.python.org/3/library/typing.html#typing.TypeVar

Reply:
Note also: this is temporary.  Python 2.12 introduces proper syntax for this.

Reply:
*shrug* agree to disagree (I'd read the documentation as structural examples), but not important enough.